### PR TITLE
Teach min-cut about online softmax prep

### LINKED ISSUE #148141: Applying online softmax patterns on joint_graph cause 1.2x peak memory regression for TB hf_T5_base model
### 🐛 Describe the bug

Command 
```time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only hf_T5_base --training```

tlparse for joint-graph pattern: https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmpk7zTmz/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=100 . Peak memory 22.2GB.

tlparse for post-grad graph pattern: https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmp9eCEvc/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=100  . Peak memory 18.6GB.

It seems like the partitioner makes worse decisions if the pattern is applied before rather than after the partitioning.

This is related to PR https://github.com/pytorch/pytorch/pull/127011 . cc @chauhang @penguinwu @jansel @eellison I'll move the pattern to post grad to avoid the regression. Ultimately we probably should improve the partitioner. cc @Chillee as well

### Error logs

_No response_

### Versions

.

cc @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185783

The online softmax preparation pattern computes softmax max and sum as a
multi-output prim. When this prim appears in an AOTAutograd joint graph, the
min-cut partitioner previously did not know that it was a recomputable/fusible
operation like the decomposed amax/sum pattern it replaces. That skewed the cut
and could make the forward save a full softmax activation instead of the much
smaller max/sum tuple outputs.

Add prims.prepare_softmax_online to the partitioner's recomputable op set. The
existing post-grad placement is still the normal path, but this makes the
partitioner robust if the online softmax preparation is introduced before
partitioning.

I considered relying only on the post-grad pattern placement that originally
avoided the regression, but that leaves the partitioner's cost model wrong for
this prim and makes future joint-graph uses fragile.

Benchmark Results:
A focused AOTAutograd min-cut reproducer that simulates joint-graph online
softmax placement saved 212,992 activation bytes before the fix and 149,504
after the fix, a 29.8% reduction. The old cut saved an extra full
[2, 128, 128] fp16 softmax activation; the new cut saves the small
[2, 128, 1] fp32 max/sum tuple outputs instead.

Test Plan:
- python test/inductor/test_online_softmax.py -k test_prepare_softmax_after_partitioning
- python test/inductor/test_online_softmax.py -k test_prepare_softmax
- python ../state/148141/benchmark_partition_saved_bytes.py
- lintrunner -a torch/_functorch/partitioners.py test/inductor/test_online_softmax.py
- git diff --check --cached

Fixes #148141
Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by shunting314 — APPROVED** [review]

**Comment by jansel** [comment]
My agent says the merge attempt was blocked by an unrelated macos-py3-arm64 MPS BatchNorm failure; I reran the failed workflow jobs to refresh CI.
